git-duet-install-hook/main.go

```diff
@@ -23,29 +30,91 @@ func main() {
 	)
 
 	getopt.Parse()
+	getopt.SetParameters(fmt.Sprintf("{ %s | %s }", preCommit, prepareCommitMsg))
 
 	if *help {
 		getopt.Usage()
 		os.Exit(0)
 	}
 
-	output := new(bytes.Buffer)
-	cmd := exec.Command("git", "rev-parse", "--show-toplevel")
-	cmd.Stdout = output
-	if err := cmd.Run(); err != nil {
+	args := getopt.Args()
+	if len(args) != 1 {
+		getopt.Usage()
+		os.Exit(1)
+	}
+	hookFileName := args[0]
+
+	var hook string
+	if hookFileName == preCommit {
+		hook = preCommitHook
+	} else if hookFileName == prepareCommitMsg {
+		hook = prepareCommitMsgHook
+	} else {
+		getopt.Usage()
+		os.Exit(1)
+	}
+
+	config, err := duet.NewConfiguration()
+	if err != nil {
 		fmt.Println(err)
 		os.Exit(1)
 	}
 
-	hookPath := path.Join(strings.TrimSpace(output.String()), ".git", "hooks", "pre-commit")
+	var hooksDir string
+	if config.Global {
+		gitConfig := &duet.GitConfig{Namespace: config.Namespace, SetUserConfig: config.SetGitUserConfig}
+		gitConfig.Scope = duet.Global
+		templateDir, err := gitConfig.GetInitTemplateDir()
+		if err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+		if templateDir == "" {
+			usr, err := user.Current()
+			if err != nil {
+				fmt.Println(err)
+				os.Exit(1)
+			}
+			templateDir = path.Join(usr.HomeDir, ".git-template")
+			if err := gitConfig.SetInitTemplateDir(templateDir); err != nil {
+				fmt.Println(err)
+				os.Exit(1)
+			}
+		}
+		if err := os.MkdirAll(path.Join(templateDir, "hooks"), os.ModePerm); err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+		hooksDir = path.Join(templateDir, "hooks")
+	} else {
+		hooksDir = getLocalHooksDir()
+	}
+
+	hookPath := path.Join(hooksDir, hookFileName)
 
-	hookFile, err := os.OpenFile(hookPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, os.ModePerm)
+	hookFile, err := os.OpenFile(hookPath, os.O_CREATE|os.O_RDWR, os.ModePerm)
 	if err != nil {
 		fmt.Println(err)
 		os.Exit(1)
 	}
 	defer hookFile.Close()
 
+	b, err := ioutil.ReadAll(hookFile)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	contents := strings.TrimSpace(string(b))
+	if contents != "" {
+		if hook == preCommitHook && contents != strings.TrimSpace(preCommitHook) ||
```

Comment:
I'm concerned that this doesn't allow for future versions of `git-duet` to overwrite with updated hook contents.

Reply:
However, this should be rare and I think we can address it if it comes up by including an index of previous hook contents for comparison (or hashes) so I'm OK with this as-is.